kong/core/balancer.lua

```diff
@@ -784,17 +784,28 @@ end
 --------------------------------------------------------------------------------
 -- Update health status and broadcast to workers
 -- @param upstream a table with upstream data
--- @param ip target IP
+-- @param hostname target hostname or IP
 -- @param port target port
 -- @param is_healthy boolean: true if healthy, false if unhealthy
 -- @return true if posting event was successful, nil+error otherwise
-local function post_health(upstream, ip, port, is_healthy)
+local function post_health(upstream, hostname, port, is_healthy)
 
   local balancer = balancers[upstream.id]
   if not balancer then
     return nil, "Upstream " .. tostring(upstream.name) .. " has no balancer"
   end
 
+  local ip
+  for weight, addr, host in balancer:addressIter() do
+    if weight > 0 and hostname == host.hostname and port == addr.port then
+      ip = addr.ip
+      break
+    end
+  end
+  if not ip then
+    return nil, "target not found for " .. hostname .. ":" .. port
+  end
```

Comment:
Reproduced locally :+1: the failure are somewhat hard to link to this particular code path (since we indirectly test this):
```
$ bin/busted spec/02-integration/05-proxy/09-balancer_spec.lua
●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●◼●●●●●●●●●●●●●●●●◼●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●
100 successes / 2 failures / 0 errors / 0 pending : 188.9688 seconds

Failure → spec/02-integration/05-proxy/09-balancer_spec.lua @ 904
Ring-balancer #postgres #localhost #healthchecks perform passive health checks -- manual recovery
spec/02-integration/05-proxy/09-balancer_spec.lua:971: Expected objects to be equal.
Passed in:
(number) 10
Expected:
(number) 20

Failure → spec/02-integration/05-proxy/09-balancer_spec.lua @ 904
Ring-balancer #cassandra #localhost #healthchecks perform passive health checks -- manual recovery
spec/02-integration/05-proxy/09-balancer_spec.lua:971: Expected objects to be equal.
Passed in:
(number) 10
Expected:
(number) 20
```

But well, I don't have a better proposal atm

Reply:
> But well, I don't have a better proposal atm

I gave this some thought, and now that we have the `/health` endpoint for introspecting status it's possible to write individual Admin API tests for the `/healthy` and `/unhealthy` endpoints. FWIW, I also added unit tests to `post_health`.